### DISCUSSION TITLE
Change from undefined to null doesn't run queryFn

### DISCUSSION THREAD

**Discussion by dmitry-seemetrics** [TanStack/query#6405] [Q&A] [answered]
React-query doesn't run queryFn if queryKey changed from undefined to null or from null to undefined.
From our backend perspective it is different things, undefined - means you didn't send value, null means you set it to null (empty)
[Example](https://codesandbox.io/p/sandbox/peaceful-panka-773dwm?file=%2Fsrc%2FApp.js%3A15%2C8&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clp6sj9x500073b6lnuz3ztjc%2522%252C%2522sizes%2522%253A%255B60.86538462%252C39.13461538%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clp6sj9x500033b6lcbb26yon%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clp6sj9x500053b6lgl4tk2b0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clp6sj9x500063b6l9qejuhoh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clp6sj9x500033b6lcbb26yon%2522%253A%257B%2522id%2522%253A%2522clp6sj9x500033b6lcbb26yon%2522%252C%2522activeTabId%2522%253A%2522clp6sm4au01003b6l7sfq4m97%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp6sj9x400023b6lx560n71z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522id%2522%253A%2522clp6sm4au01003b6l7sfq4m97%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522clp6smyrm01ea3b6l3cfr7krn%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clp6sj9x500063b6l9qejuhoh%2522%253A%257B%2522id%2522%253A%2522clp6sj9x500063b6l9qejuhoh%2522%252C%2522activeTabId%2522%253A%2522clp6sqfi502ol3b6lgmvfr5bz%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clp6sqfi502ol3b6lgmvfr5bz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clp6sj9x500053b6lgl4tk2b0%2522%253A%257B%2522id%2522%253A%2522clp6sj9x500053b6lgl4tk2b0%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp6sj9x500043b6lctqthw8v%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clp6sja2n0002dier849zci5z%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clp6sj9x500043b6lctqthw8v%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
What can I do if my queryKey needs to change fron undefined to null for example?

**Comment by TkDodo** [comment, marked as answer]
Our internal query key hashing is based on `JSON.stringify`, which will yield `'[null]'` for both `[null]` and `[undefined]`.

If that's an issue for you, you can provide your own `queryKeyHashFn`. You can also do that globally as a default option.

The docs say:

> When a query needs more information to uniquely describe its data, you can use an array with a string and any number of serializable objects to describe it. This is useful for:

but I agree that could be described a bit better. If you want, please file a PR to update the docs?

https://tanstack.com/query/latest/docs/react/guides/query-keys#query-keys-are-hashed-deterministically

**Comment by wpp-roman-gryndii** [comment]
@TkDodo I have another issue related to hashing

I have an endpoint /messages?maxIndex=number|undefined

I am changing this parameter in the key dynamically from
['QUERY_KEY', { parameter: 777 }] > ['QUERY_KEY', { parameter: undefined }]

And then call query.refetch() it is being re-fetched with the previous value 777 instead of undefined

is that expected?

**Reply by TkDodo** [reply]
show a reproduction please, might be related to:

- https://github.com/TanStack/query/issues/3741
